### PR TITLE
Add draft one pager for Provider

### PR DESCRIPTION
**Pull request by rahulgrover99**
Adds a draft one-pager for the provider-nop.

Signed-off-by: Rahul Grover <rahulgrover99@gmail.com>

### CODE REVIEW

File: design/one-pager-provider-nop..md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+* Reviewers: Crossplane Maintainers
+* Status: Speculative
+
+## Introduction
+
+There is a lot of machinery around the composition engine that needs to be tested which raises demand for a provider on which could rely to behave the way we want it to. It can be used to mock the behavior of providers by creating fake objects whenever needed so that we could test the response of composition to the resource becoming ready or not. 
+
+This provider does exactly what you tell it to. It can be used for testing functionality in core Crossplane without needing to account for the requirements of external APIs. It supports a single managed resource that contains a spec that allows for specifying how and when the status of the resource is changed. It does not orchestrate any external system. 
+
+This provider will be installed while running Crossplane end-to-end tests to satisfy any kind of composite infrastructure resource by doing nothing to simulate different scenarios.
```
Comment (jbw976) [inline design/one-pager-provider-nop..md]: "doing nothing" makes it sound like it will literally not do anything at all.  perhaps reword this to capture:

* The controller for this nop type will not call any external APIs during its reconcile loop
* The controller will take action to perform the behavior that the user has declared in its spec

File: design/one-pager-provider-nop..md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+## Design
+
+The provider will have a single managed resource `NopResource` that will reflect whatever is in the spec into the status. It will let you define to:
+- become ready after this period
+- become unhealthy after this period
+
+The config might look something like:
+
+```yaml
+spec:
```
Comment (jbw976) [inline design/one-pager-provider-nop..md:75]: looks like this is the quick draft spec we came up with in our meeting, but i think it needs a bit more refinement.

take a look at the crossplane-runtime common conditions in https://github.com/crossplane/crossplane-runtime/blob/master/apis/common/v1/condition.go#L205-L214

what would we change about this proposed nop objet spec to be more compatible with that common condition format that all crossplane resources use?

File: design/one-pager-provider-nop..md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Crossplane provider-nop
+
+* Owner: Rahul Grover (@rahulgrover99)
+* Reviewers: Crossplane Maintainers
+* Status: Speculative
+
+## Introduction
```
Comment (jbw976) [inline design/one-pager-provider-nop..md:7]: i think we could use an anchoring scenario to give this doc a better foundation, it feels a bit too hypothetical right now.

let's try to describe a real test scenario that we would use this provider-nop and its objects for.  maybe something like:

* test case creates an XRD and a composition of 3 nop objects
  * each of those nop objects is told to become ready after 3 seconds
* test case creates an XR instance of this XRD
* test waits and verifies that 3 nops objects are created, they all make it to the ready status, and the XR itself then makes it to the ready status because all 3 of its composed resources are now ready.

File: design/one-pager-provider-nop..md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+- become unhealthy after this period
+
+The config might look something like:
+
+```yaml
+spec:
+    condtionAfter:
+        time: 5s
+        condition: ready
+```
```
Comment (jbw976) [inline design/one-pager-provider-nop..md:84]: another field in the spec that I can imagine would be a receiver field to test that patches make it into the nop object when it's part of a composition.  e.g.:

* nop object spec has a field called 'patchReceiverField` of type string
* XRD has a field called `coolField` of type string
* The composition under the XRD creates 1 nop object and patches the value from `coolField` onto this receiver field on the nop object.
* test case verifies that the value of `coolField` made it successfully into the nop object's receiver field

**Review by jbw976 — COMMENTED** [review]
thanks for driving this design @rahulgrover99! i've left a few suggestions to take this design a little deeper, let me know if you have any questions and we can sync up about the feedback!

File: design/one-pager-provider-nop..md (hunk 2 above)
Comment (rahulgrover99) [reply design/one-pager-provider-nop..md:75]: Yes, actually this is just a draft I wrote to check if I was thinking on the right path. Now that I think of the crossplane-runtime common conditions, will it work if we take these different conditions as a string type and then implement a switch case logic in the `Create` method of the controller?

File: design/one-pager-provider-nop..md (hunk 3 above)
Comment (rahulgrover99) [reply design/one-pager-provider-nop..md:7]: Sure, I will add such example to `Introduction` 👍🏼

File: design/one-pager-provider-nop..md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  forProvider:
+    conditionAfter:
+      - conditionType: "Synced"
+        conditionStatus: "True"
+        timeAfter: "5s"
+      - conditionType: "Ready"
+        conditionStatus: "False"
+        timeAfter: "10s"
+```
+`ObservableField` is an arbitrary field in `NopResourceObservation` type can be useful in crossplane testing scenarios. For example, while testing bidirectional patching back to the composite resource from one resource to another which can be supported by this field.
```
Comment (jbw976) [inline design/one-pager-provider-nop..md]: I don't see `observableField` shown in the schema above, is this paragraph stale?

File: design/one-pager-provider-nop..md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+#### Connect
+Currently, there is no requirement for having `ProviderConfig` since we don't have any external API client that we might want to connect. 
+We would just create an external client with no service parameter passed. 
+
+#### Observe
+This is where the main logic will reside that would compare the time elapsed with the time interval passed in the spec. As the time elapsed becomes equal to the time passed it will change the condition of resource accordingly. 
+
+#### Update
+Currently there is no plan to update the NopResource so this function will just set condition to crossplane-runtime `Available`.
```
Comment (jbw976) [inline design/one-pager-provider-nop..md]: This doesn't sound correct to me, because it would likely be in conflict with the status conditions we are supposed to be setting according to the desired behavior from the spec.  e.g. if spec says ready: false after 5 seconds, we wouldn't want the Update function to set it to Available, that would effectively wipe out the ready: false that we desire.

One potential flow that may be aligned with the design of the managed reconciler would be:

1) Observe function should determine if the status conditions match what the spec says they should be.  If they do not match (i.e. we need to update the status conditions), it can return `ResourceUpToDate: false`, indicating that Update should be called to perform the update.  See more about this value in https://github.com/crossplane/provider-template/blob/master/internal/controller/mytype/mytype.go#L145-L148
2) Update function should then carry out the intended status/conditions change as per what the spec is declaring.  It makes more sense in my mind that updating the resource/status should be done in the Update function.

Thoughts?

File: design/one-pager-provider-nop..md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      - conditionType: "Ready"
+        conditionStatus: "False"
+        timeAfter: "10s"
+```
+`ObservableField` is an arbitrary field in `NopResourceObservation` type can be useful in crossplane testing scenarios. For example, while testing bidirectional patching back to the composite resource from one resource to another which can be supported by this field. 
+
+The idea here is just to wait for the time provided in the spec before allowing the resource to be ready/unhealthy. This could be achieved by adding the logic to the controllers. 
+
+## Future Plans
+- We might want to include `ProviderConfig` at some point for those resources which are not managed.
```
Comment (jbw976) [inline design/one-pager-provider-nop..md]: can you explain this thought a little more? what resources are not managed?

File: design/one-pager-provider-nop..md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        timeAfter: "10s"
+```
+`ObservableField` is an arbitrary field in `NopResourceObservation` type can be useful in crossplane testing scenarios. For example, while testing bidirectional patching back to the composite resource from one resource to another which can be supported by this field. 
+
+The idea here is just to wait for the time provided in the spec before allowing the resource to be ready/unhealthy. This could be achieved by adding the logic to the controllers. 
+
+## Future Plans
+- We might want to include `ProviderConfig` at some point for those resources which are not managed. 
+- Array type in nop resource observation type like `ObservableArrays []string` can also be useful in patching.
+- A `patchReceiverField` can added as an optional field in `NopResource` parameters that just gets patched value from `coolField` of XRD
```
Comment (jbw976) [inline design/one-pager-provider-nop..md]: maybe add a bit more detail to these last 2 future plan items too: e.g. explain WHY they would be useful and what they would help test.

**Review by jbw976 — COMMENTED** [review]
This is looking good, I see most all of our sync-up yesterday incorporated into the design.  A few more comments and behavior to clarify :)

File: design/one-pager-provider-nop..md (hunk 5 above)
Comment (rahulgrover99) [reply design/one-pager-provider-nop..md]: Actually, I had in mind that this field will be a part of `NopResourceObservation` and not `NopResourceParameters` making it part of spec under `atProvider` and not `forProvider`. This field would be displayed in the status. For our use case, it might act as a placeholder for storing some fake information. This will allow us to test something related to status, like @hasheddan mentioned about bidirectional patching in the sync-up call (status going from bucket to composite and then propagating to another resource).

File: design/one-pager-provider-nop..md (hunk 6 above)
Comment (rahulgrover99) [reply design/one-pager-provider-nop..md]: Yes, that makes perfect sense to me @jbw976. I also was having second thoughts about this and hence thought to jot it down so that we could discuss it. Let's go with the above-mentioned flow 👍🏼

File: design/one-pager-provider-nop..md (hunk 7 above)
Comment (rahulgrover99) [reply design/one-pager-provider-nop..md]: Yeah, I realize I might have put it in the wrong way. I was trying to point out that there might be an edge case where we want to test with `ProviderConfig`. I guess @muvaf was mentioning something about it that I might have misinterpreted.

**Review by jbw976 — APPROVED** [review]
Thanks for all the efforts on this @rahulgrover99!
